File: src/Juniper.TSLib/src/units/fileSize.ts
type Base = 2 | 10;

const labels = new Map<Base, Map<number, string>>([
    [2, new Map<number, string>([
        [0, "B"],
        [1, "KiB"],
        [2, "MiB"],
        [3, "GiB"],
        [4, "TiB"]
    ])],
    [10, new Map<number, string>([
        [0, "B"],
        [1, "KB"],
        [2, "MB"],
        [3, "GB"],
        [4, "TB"]
    ])]
]);

export function formatBytes(value: number, base: 2 | 10 = 10) {
    const isNegative = value < 0;
    value = Math.abs(value);

    const systemBase = base === 2 ? 1024 : 1000;
    let size = Math.min(4, Math.floor(Math.log(value) / Math.log(systemBase)));
    let divisor = Math.pow(systemBase, size);
    if (2 * value >= systemBase * divisor && size < 4) {
        size++;
        divisor *= systemBase;
    }
    const levels = labels.get(base);
    const label = levels.get(size);
    const converted = value / divisor;
    const str = `${isNegative ? "-" : ""}${converted.toFixed(2)} ${label}`;

    return str;
}

File: src/Juniper.Root/Units/FileSize.cs
using Accord.Math;

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Juniper.Units
{
    public static class FileSize
    {
        private static readonly Dictionary<SystemOfMeasure, UnitOfMeasure[]> systemLevels = new Dictionary<SystemOfMeasure, UnitOfMeasure[]>(2)
        {
            [SystemOfMeasure.Metric] = new[] {
                UnitOfMeasure.Bytes,
                UnitOfMeasure.Kilobytes,
                UnitOfMeasure.Megabytes,
                UnitOfMeasure.Gigabytes,
                UnitOfMeasure.Terabytes,
                UnitOfMeasure.Petabytes,
                UnitOfMeasure.Exabytes,
                UnitOfMeasure.Zettabytes,
                UnitOfMeasure.Yotabytes
            },
            [SystemOfMeasure.USCustomary] = new[] {
                UnitOfMeasure.Bytes,
                UnitOfMeasure.Kibibytes,
                UnitOfMeasure.Mibibytes,
                UnitOfMeasure.Gibibytes,
                UnitOfMeasure.Tebibytes,
                UnitOfMeasure.Pebibytes,
                UnitOfMeasure.Exbibytes,
                UnitOfMeasure.Zebibytes,
                UnitOfMeasure.Yobibytes
            }
        };

        public static string Format(long value, UnitOfMeasure units = UnitOfMeasure.Bytes)
        {
            var isNegative = value < 0;
            value = Math.Abs(value);
            var str = Format((ulong)value, units);
            if (isNegative)
            {
                str = System.Globalization.NumberFormatInfo.CurrentInfo.NegativeSign + str;
            }

            return str;
        }

        public static string Format(ulong value, UnitOfMeasure units = UnitOfMeasure.Bytes)
        {
            if (value < 1)
            {
                return "0";
            }
            else
            {
                var system = (from kv in systemLevels
                              where kv.Value.Contains(units)
                              select (SystemOfMeasure?)kv.Key)
                            .FirstOrDefault();
                if (system is null)
                {
                    throw new ArgumentException("Units parameter must be a file type. Given: " + units, nameof(units));
                }

                var systemBase = system == SystemOfMeasure.Metric ? 1000 : 1024;
                var deltaSize = (int)(Math.Log(value) / Math.Log(systemBase));
                var divisor = Math.Pow(systemBase, deltaSize);
                if (2 * value >= systemBase * divisor)
                {
                    deltaSize += 1;
                }

                var levels = systemLevels[system.Value];
                var curSize = levels.IndexOf(units);
                var endSize = Math.Min(curSize + deltaSize, levels.Length - 1);
                var convertTo = levels[endSize];

                return Converter.Label(value, units, convertTo);
            }
        }
    }
}
